src/platform/plugins/shared/workflows_execution_engine/server/step/custom_step_impl.ts

```diff
@@ -75,19 +75,26 @@ export class CustomStepImpl extends BaseAtomicNodeImplementation<BaseStep> {
   private createHandlerContext(input: unknown): StepHandlerContext {
     return {
       input,
+      rawInput: this.node.configuration.with || {},
       contextManager: {
         getContext: () => {
           return this.stepExecutionRuntime.contextManager.getContext();
         },
         getScopedEsClient: () => {
           return this.stepExecutionRuntime.contextManager.getEsClientAsUser();
         },
-        renderInputTemplate: (value) => {
-          return this.stepExecutionRuntime.contextManager.renderValueAccordingToContext(value);
+        renderInputTemplate: (value, additionalContext) => {
+          return this.stepExecutionRuntime.contextManager.renderValueAccordingToContext(
+            value,
+            additionalContext
+          );
         },
         getFakeRequest: () => {
           return this.stepExecutionRuntime.contextManager.getFakeRequest();
         },
+        setVariables: (variables: Record<string, unknown>) => {
```

Comment:
Do we want to allow custom steps to define some variables?
I imagined custom steps as simple action steps.

Reply:
I tend to think of data.* steps as core steps that belong to the execution engine. Hopefully that makes sense.
If we expose setVariables API externally, do we expect others to be able to allow setting variables besides data.setVariable?

Reply:
> wdym @skynetigor?

I mean that each step that will use setVariables will need to specify variables schema it will assign so editor can understand what exactly variables are available during suggestions/completions.

Reply:
I don't understand very well the goal of the `variables`, I don't think they hurt, but I am not sure they are necessary: 
If it's just a way to store information for other steps to access, what is the difference with the step `output`?

Reply:
> I mean that each step that will use setVariables will need to specify variables schema it will assign so editor can understand what exactly variables are available during suggestions/completions.

@skynetigor but we already generically do that..

> If it's just a way to store information for other steps to access, what is the difference with the step output?

Just some sugar syntax, no real difference. It's not really the "output" of the step, but "variables" that you set on the context of the workflow.